### PR TITLE
LibCrypto: Add ChaCha20

### PR DESCRIPTION
**Pull request by stelar7**
Part 2/3 of adding ChaCha20Poly1305 to LibTLS

### CODE REVIEW

File: Userland/Libraries/LibCrypto/Cipher/ChaCha20.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace Crypto::Cipher {
+
+class ChaCha20 {
+    static constexpr u32 CONSTANT_16_BYTES[] { 0x61707865, 0x3120646E, 0x79622D36, 0x6B206574 };
+    static constexpr u32 CONSTANT_32_BYTES[] { 0x61707865, 0x3320646E, 0x79622D32, 0x6B206574 };
+
+public:
+    ChaCha20(ReadonlyBytes key, ReadonlyBytes nonce, u32 initial_counter = 0);
+    ErrorOr<ByteBuffer> run_cipher(ReadonlyBytes input);
```
Comment (alimpfard) [inline Userland/Libraries/LibCrypto/Cipher/ChaCha20.h:19]: Let's at least keep the interface somewhat consistent and add encrypt()/decrypt()
It'd also be nice if the main impl took a `Bytes output`, I believe the code in TLS expects something like that (have you checked if any TLS bits can be reused for this?)

Comment (stelar7) [reply Userland/Libraries/LibCrypto/Cipher/ChaCha20.h:19]: Havent looked at anything tls yet, thats step 3 :^)

File: Userland/Libraries/LibCrypto/Cipher/ChaCha20.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        }
+
+        // XOR the input and the current block
+        u32 n = min(input.size() - offset, 64 - m_state.block_offset);
+        u8* key_block = (u8*)m_state.block + m_state.block_offset;
+        for (u32 i = 0; i < n; i++) {
+            u8 input_byte = input.offset_pointer(offset)[i];
+            u8 key_byte = key_block[i];
+
+            ByteReader::store(output.offset_pointer(offset + i), input_byte ^ key_byte);
```
Comment (ADKaster) [inline Userland/Libraries/LibCrypto/Cipher/ChaCha20.cpp:139]: ASAN is complaining here that your offset + i is exactly the same as the sizeof(output), and you're writing one outside the bounds of the ByteBuffer you created in the test.

File: Userland/Libraries/LibCrypto/Cipher/ChaCha20.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+private:
+    void run_cipher(ReadonlyBytes input, Bytes& output);
+    void process_block();
+    ALWAYS_INLINE void do_quarter_round(u32& a, u32& b, u32& c, u32& d);
+
+    struct State {
+        u32 state[16];
+        u32 block[16];
+        u8 block_offset;
+    } m_state;
```
Comment (IdanHo) [inline Userland/Libraries/LibCrypto/Cipher/ChaCha20.h:32]: Explicitly initialize these? I think the usage of `m_state.block_offset` above depends on uninitialized values.

File: Userland/Libraries/LibCrypto/Cipher/ChaCha20.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+private:
+    void run_cipher(ReadonlyBytes input, Bytes& output);
+    void process_block();
+    ALWAYS_INLINE void do_quarter_round(u32& a, u32& b, u32& c, u32& d);
+
+    struct State {
+        u32 state[16] {};
+        u32 block[16] {};
+        u8 block_offset {};
+    } m_state;
```
[the remark this answers is not in the document]
Comment (stelar7) [reply Userland/Libraries/LibCrypto/Cipher/ChaCha20.h]: There was no real point in making it a struct, I just started with that and didnt think to unpack it 😅 
Did that now 👍

**Pull request merged**